lib/LanguageServerWorseReflection/Workspace/WorkspaceIndex.php

```diff
@@ -37,41 +45,64 @@ public function documentForName(Name $name): ?TextDocument
 
     public function index(TextDocument $textDocument): void
     {
+        $this->documents[(string)$textDocument->uri()] = $textDocument;
+        $this->updateDocument($textDocument);
+    }
+
+    private function updateDocument(TextDocument $textDocument): void
+    {
+        $newNames = [];
         foreach ($this->reflector->reflectClassesIn($textDocument) as $reflectionClass) {
-            $this->byName[$reflectionClass->name()->full()] = $textDocument;
+            $newNames[] = $reflectionClass->name()->full();
         }
-
+        
         foreach ($this->reflector->reflectFunctionsIn($textDocument) as $reflectionFunction) {
-            $this->byName[$reflectionFunction->name()->full()] = $textDocument;
+            $newNames[] = $reflectionFunction->name()->full();
+        }
+
+        $this->updateNames($textDocument, $newNames, $this->documentToNameMap[(string)$textDocument->uri()] ?? []);
+    }
+
+    private function updateNames(TextDocument $textDocument, array $newNames, array $currentNames): void
+    {
+        $namesToRemove = array_diff($currentNames, $newNames);
+        
+        foreach ($newNames as $name) {
+            $this->byName[$name] = $textDocument;
+        }
+        foreach ($namesToRemove as $name) {
+            unset($this->byName[$name]);
+        }
+
+        if (!empty($newNames)) {
+            $this->documentToNameMap[(string)$textDocument->uri()] = $newNames;
+        } else {
```

Comment:
minor but I would remove else in favor of an early return:

```
if ($something) {
    // do something
    return;
}

// do something else
```

Reply:
Sure, but I am not sure how to append this to the current PS - push to the master branch again?

Reply:
It's fine, it's merged now.